src/DataStructures/Tensor/Expressions/Evaluate.hpp

```diff
@@ -38,8 +40,158 @@ constexpr bool contains_indices_to_contract(
+/// The elements of the transformation are the positions of the RHS generic
+/// indices in the LHS generic indices. Put another way, for some `i`,
+/// `rhs_tensorindices[i] == lhs_tensorindices[index_transformation[i]]`.
+///
+/// Here is an example of what the algorithm does:
+///
+/// Tensor equation: \f$L_{cab} = R_{abc}\f$
+/// `lhs_tensorindices`:
+/// \code
+/// {2, 0, 1} // i.e. {c, a, b}
+/// \endcode
+/// `rhs_tensorindices`:
+/// \code
+/// {0, 1, 2} // i.e. {a, b, c}
+/// \endcode
+/// returned `index_transformation`:
+/// \code
+/// {1, 2, 0} // positions of RHS indices {c, a, b} in LHS indices {a, b, c}
+/// \endcode
+///
+/// \tparam NumIndices the number of indices in the tensors
+/// \param lhs_tensorindices the TensorIndexs of the LHS tensor
+/// \param rhs_tensorindices the TensorIndexs of the RHS tensor
+/// \return a transformation from the LHS tensor's multi-indices to the
+/// equivalent RHS tensor's multi-indices
+template <size_t NumIndices>
+SPECTRE_ALWAYS_INLINE constexpr std::array<size_t, NumIndices>
+compute_index_transformation(
+    const std::array<size_t, NumIndices>& lhs_tensorindices,
+    const std::array<size_t, NumIndices>& rhs_tensorindices) noexcept {
+  std::array<size_t, NumIndices> index_transformation{};
+  for (size_t i = 0; i < NumIndices; i++) {
+    gsl::at(index_transformation, i) = static_cast<size_t>(std::distance(
+        lhs_tensorindices.begin(),
+        alg::find(lhs_tensorindices, gsl::at(rhs_tensorindices, i))));
+  }
+  return index_transformation;
+}
+
+/// \brief Computes the RHS tensor multi-index that is equivalent to a given
+/// LHS tensor multi-index, according to the differences in the orderings of
+/// their generic indices
+///
+/// \details
+/// Here is an example of what the algorithm does:
+///
+/// Tensor equation: \f$L_{cab} = R_{abc}\f$
+/// `index_transformation`:
+/// \code
+/// {1, 2, 0} // positions of RHS indices {c, a, b} in LHS indices {a, b, c}
+/// \endcode
+/// `lhs_multi_index`:
+/// \code
+/// {3, 4, 5} // i.e. c = 3, a = 4, b = 5
+/// \endcode
+/// returned equivalent `rhs_multi_index`:
+/// \code
+/// {4, 5, 3} // i.e. a = 4, b = 5, c = 3
+/// \endcode
+///
+/// \tparam NumIndices the number of indices in the tensors
+/// \param lhs_multi_index the multi-index of the LHS tensor
+/// \param index_transformation the list of the positions of the RHS indices in
+/// the LHS indices
+/// \return the RHS tensor multi-index that is equivalent to `lhs_tensor_index`
+template <size_t NumIndices>
+SPECTRE_ALWAYS_INLINE constexpr std::array<size_t, NumIndices>
+compute_rhs_multi_index(
+    const std::array<size_t, NumIndices>& lhs_multi_index,
+    const std::array<size_t, NumIndices>& index_transformation) noexcept {
+  std::array<size_t, NumIndices> rhs_multi_index{};
+  for (size_t i = 0; i < NumIndices; i++) {
+    gsl::at(rhs_multi_index, i) =
+        gsl::at(lhs_multi_index, gsl::at(index_transformation, i));
+  }
+  return rhs_multi_index;
+}
+
```

Comment:
Note that the `compute_index_transformation` and `compute_rhs_multi_index` functions added here are directly copy + pasted from `TensorAsExpression.hpp`([link](https://github.com/sxs-collaboration/spectre/blob/63ddd991fc2f18674a8e93b14057c691278d973f/src/DataStructures/Tensor/Expressions/TensorAsExpression.hpp#L89)).  I think it may be a good idea to factor these two functions out into another file that holds general purpose functions (name suggestions welcome!) that are useful for implementing more than one file or expression type. I didn't do that refactoring in this PR because I feel the PR is already large enough. Looking for opinions on this.

Reply:
I've factored these out on a local branch and decided to go with a slightly longer name that is very similar in spirit to your suggestions: TensorIndexTransformation.hpp